### PR TITLE
feat(core): generateSignInUri

### PR DESCRIPTION
**Pull request by xiaoyijun**
## Summary

* feat(core): generateSignInUri

## Linear Issue Reference

* LOG-748

## Testing

UT

@logto-io/eng

### CODE REVIEW

File: kotlin/src/main/kotlin/io/logto/sdk/core/Core.kt [hunk 1]
```diff
@@ -1,10 +1,40 @@
...
+        return constructedUri.newBuilder().apply {
+            addQueryParameter(QueryKey.CLIENT_ID, clientId)
+            addQueryParameter(QueryKey.CODE_CHALLENGE, codeChallenge)
+            addQueryParameter(QueryKey.CODE_CHALLENGE_METHOD, CodeChallengeMethod.S256)
+            addQueryParameter(QueryKey.STATE, state)
+            addQueryParameter(QueryKey.REDIRECT_URI, redirectUri)
+            addQueryParameter(QueryKey.PROMPT, PromptValue.CONSENT)
+            addQueryParameter(QueryKey.RESPONSE_TYPE, ResponseType.CODE)
+            addQueryParameter(QueryKey.SCOPE, scope.ensureDefaultScopes().joinToString(" "))
+            resource?.let { for (resVal in it) { addQueryParameter(QueryKey.RESOURCE, resVal) } }
```
Comment (gao-sun) [inline kotlin/src/main/kotlin/io/logto/sdk/core/Core.kt:34]:
```suggestion
            resource?.let { for (value in it) { addQueryParameter(QueryKey.RESOURCE, value) } }
```

File: kotlin/src/main/kotlin/io/logto/sdk/core/constant/ReservedScope.kt [hunk 2]
```diff
@@ -0,0 +1,6 @@
+package io.logto.sdk.core.constant
+
+object ReservedScope {
+    const val OPEN_ID = "openid"
```
Comment (gao-sun) [inline kotlin/src/main/kotlin/io/logto/sdk/core/constant/ReservedScope.kt:4]: maybe
```suggestion
    const val OPENID = "openid"
```

File: kotlin/src/main/kotlin/io/logto/sdk/core/extension/StringListExt.kt [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+    if (!mutableList.contains(ReservedScope.OPEN_ID)) {
+        mutableList.add(ReservedScope.OPEN_ID)
+    }
+
+    if (!mutableList.contains(ReservedScope.OFFLINE_ACCESS)) {
+        mutableList.add(ReservedScope.OFFLINE_ACCESS)
+    }
+
+    return mutableList.toList()
+} ?: listOf(ReservedScope.OPEN_ID, ReservedScope.OFFLINE_ACCESS)
```
Comment (gao-sun) [inline kotlin/src/main/kotlin/io/logto/sdk/core/extension/StringListExt.kt:21]: maybe something like:
```kt
((this ?: listOf()) + listOf(ReservedScope.OPEN_ID, ReservedScope.OFFLINE_ACCESS)).distinct()
```

**Review by gao-sun — CHANGES_REQUESTED** [review]

File: kotlin/src/main/kotlin/io/logto/sdk/core/extension/StringListExt.kt [hunk 4]
```diff
@@ -0,0 +1,21 @@
+package io.logto.sdk.core.extension
+
+import io.logto.sdk.core.constant.ReservedScope
+
+fun List<String>?.ensureDefaultScopes(): List<String> = this?.let {
+    if (it.contains(ReservedScope.OPEN_ID) and it.contains(ReservedScope.OFFLINE_ACCESS)) {
+        return it
+    }
+
+    val mutableList = it.toMutableList()
```
Comment (wangsijie) [inline kotlin/src/main/kotlin/io/logto/sdk/core/extension/StringListExt.kt:10]: what about a more specific name like `scopes`?

Comment (xiaoyijun) [reply kotlin/src/main/kotlin/io/logto/sdk/core/extension/StringListExt.kt]: updated

File: kotlin/src/main/kotlin/io/logto/sdk/core/extension/StringListExt.kt (hunk 3 above)
Comment (xiaoyijun) [reply kotlin/src/main/kotlin/io/logto/sdk/core/extension/StringListExt.kt]: updated

File: kotlin/src/main/kotlin/io/logto/sdk/core/Core.kt (hunk 1 above)
Comment (xiaoyijun) [reply kotlin/src/main/kotlin/io/logto/sdk/core/Core.kt]: updated

File: kotlin/src/main/kotlin/io/logto/sdk/core/constant/ReservedScope.kt (hunk 2 above)
Comment (xiaoyijun) [reply kotlin/src/main/kotlin/io/logto/sdk/core/constant/ReservedScope.kt]: updated

**Review by gao-sun — APPROVED** [review]
lgtm

**Pull request merged**